app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,322 @@
+    }
+
+    /// End the profiling of the named interval.
+    end(interval_name: string) {
+        let start_label = this.startIntervalLabel(interval_name)
+        let end_label = this.endIntervalLabel(interval_name)
+        let measurement_label = this.measureIntervalLabel(interval_name)
+
+        if (this.isActive()) {
+            performance.mark(end_label)
+            performance.measure(measurement_label, start_label, end_label)
+        }
+    }
+
+    /// Profile the execution of the given callable.
+    measure(interval_name: string, closure: CallableFunction) {
+        this.start(interval_name)
+        const ret = closure()
+        this.end(interval_name)
+        return ret
+    }
+}
+
+// =========================
+// === Encoding/Decoding ===
+// =========================
+
+/// Delimiter used to to encode information in the `PerformanceEntry` name.
+const MESSAGE_DELIMITER = '//'
+
+/**
+ * Encodes the given information in a string. This is done by separating the information by the `MESSAGE_DELIMITER`.
+ * Example output: "TASK//SomeWork//start"
+ */
+function encodeLabel(log_level: Profiler, inter_val_name: string, event: string): string {
+    return [log_level.log_level, inter_val_name, event].join(MESSAGE_DELIMITER)
+}
+
+function parseLogLevel(log_level_name: string): Profiler | undefined {
+    for (const profiler of profiler_registry) {
+        if (profiler.isLogLevelName(log_level_name)) {
+            return profiler
+        }
+    }
+    return undefined
+}
+
+/**
+ * Decodes the given information in a string that was encoded with `encodeLabel`.
+ */
+function decodeLabel(label: string): [Profiler, string] | undefined {
+    let parts = label.split(MESSAGE_DELIMITER)
+    try {
+        let [log_level_name, interval_name] = [parts[0], parts[1]]
+        let profiler = parseLogLevel(log_level_name)
+        return [profiler, interval_name]
+    } catch (e) {
+        return undefined
+    }
+}
+
+// ======================
+// === IntervalHandle ===
+// ======================
+
+/**
+ * Handle that allows the setting of an `end` marker for a profiling interval that was started via
+ * the `start` function.
+ */
+class IntervalHandle {
+    interval_name: string
+    log_level: Profiler
+    released_status: ReleasedStatus
+
+    constructor(interval_name: string, log_level: Profiler) {
+        this.interval_name = interval_name
+        this.log_level = log_level
+        this.released_status = new ReleasedStatus()
+        this.released_status.released = false
+        handle_registry.register(
+            this,
+            new IntervalFinalizationRegistryEntry(interval_name, this.released_status)
+        )
+    }
+
+    /// Measure the interval.
+    end() {
+        this.log_level.end(this.interval_name)
+        this.released_status.released = true
+    }
+
+    /// Release the handle to manually call `end_interval` without emitting a warning.
+    release() {
+        this.released_status.released = true
```

Comment:
code repetition + should `release` not be named `drop`? Maybe not, just asking.

Reply:
This is not the same as `drop` in rust. Instead, if allows one to manually opt out of the tracking of whether the handle is "dropped" (for example because the other end of the interval is in Rust code). There is an equivalent `release` method in the Rust API.  I have clarified the documentation here to make this more clear.